stan-dev/math
# Fix the new TBB interface and use tbb::task_arena

## Summary

This builds on https://github.com/stan-dev/math/pull/2257 and fixes the related tests when the TBB environment variables are set correctly to use an externally updated library, including [oneTBB](https://github.com/oneapi-src/oneTBB). It fixes the `LDFLAGS_TBB` compiler flags and uses `tbb::task_arena` in replacement of the old (and now removed) `tbb::task_schedular_init`.

- [X] Using `tbb::task_arena` and `tbb::global_control` to manage the task scheduler arena.
- [X] Fixed compilation errors with when linking external Intel  [oneTBB](https://github.com/oneapi-src/oneTBB) / [OneAPI](https://software.intel.com/content/www/us/en/develop/articles/intel-oneapi-toolkit-release-notes.html).
- [X] Fixed global control to restrict the maximum allowed number of threads.
- [X] Updated compiler flags to fix unit tests with `oneTBB`.
- [X] Updated Intel TBB instructions in the README.

This is related to https://github.com/stan-dev/math/pull/2257 and https://github.com/RcppCore/RcppParallel/pull/141.

## Tests


- Install [oneTBB](https://github.com/oneapi-src/oneTBB) or [OneAPI](https://software.intel.com/content/www/us/en/develop/articles/intel-oneapi-toolkit-release-notes.html).

For example, installing [oneTBB](https://github.com/oneapi-src/oneTBB) on Linux 64-bit (`x86_64`) to `$HOME` directory (change if needed!):
```
    TBB_VERSION="2021.1.1"

    wget https://github.com/oneapi-src/oneTBB/releases/download/v2021.1.1/oneapi-tbb-$TBB_VERSION-lin.tgz
    tar zxvf oneapi-tbb-$TBB_VERSION-lin.tgz -C $HOME

    export TBB="$HOME/oneapi-tbb-$TBB_VERSION"
```

- Set the TBB environment variables (specifically: `TBB` for the installation prefix, `TBB_INC` for the directory that includes the header files, and `TBB_LIB` for the libraries directory).

For example, installing [oneTBB](https://github.com/oneapi-src/oneTBB) on Linux 64-bit (`x86_64`) to `$HOME` directory (change if needed!):
```
    source $TBB/env/vars.sh intel64

    export TBB_INC="$TBB/include"
    export TBB_LIB="$TBB/lib/intel64/gcc4.8"
    #export LD_LIBRARY_PATH="$TBB_LIB:$LD_LIBRARY_PATH"

    mkdir -p ~/.config/stan
    echo TBB_INTERFACE_NEW=true>> ~/.config/stan/make.local

    # Checks:
    ls -lAh $TBB_INC
    # should list `oneapi` and `tbb` directories that include the headers
    ls -lAh $TBB_LIB
    # should list the TBB libraries, including `libtbb.so`
```

- Run TBB tests:
```
> ./runTests.py test/unit/math/prim/core/init_threadpool_tbb_test.cpp
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from intel_tbb_new_init
[ RUN      ] intel_tbb_new_init.check_status
[       OK ] intel_tbb_new_init.check_status (1 ms)
[----------] 1 test from intel_tbb_new_init (1 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (2 ms total)
[  PASSED  ] 1 test.
```

```
> ./runTests.py test/unit/math/prim/core/init_threadpool_tbb_late_test.cpp
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from intel_tbb_new_late_init
[ RUN      ] intel_tbb_new_late_init.check_status
[       OK ] intel_tbb_new_late_init.check_status (0 ms)
[----------] 1 test from intel_tbb_new_late_init (0 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (0 ms total)
[  PASSED  ] 1 test.
```

Note that the test names for the new TBB interface include a `tbb_new_` prefix while the default tests use `tbb_` prefix. So, make sure that you run the new tests (`intel_tbb_new_init` and `intel_tbb_new_late_init` reported in the test results) or otherwise your build has used the internal TBB source code (i.e., the environment isn't set correctly for using the external TBB library).

## Side Effects

N/A

## Release notes

Added support for the new TBB interface and allowed using an external TBB library.

## Checklist

- [x] Math issue N/A

- [x] Copyright holder: Hamada S. Badr <hamada.s.badr@gmail.com>

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested
